Bluetooth: Controller: Fix Tx ISO SDUs dropped due to strict SN check

cvinayak opened the pull request:
Introduce a Kconfig BT_CTLR_ISOAL_SN_STRICT to relax the Tx
ISO Data SDUs being dropped due to strict check of sequence
numbers in ISOAL, i.e. dropped when ISO Data SDUs are
delayed from upper layer with respect to the current
payload number in the ISO radio events.

Relax Tx ISO SDUs being dropped in unicast audio client and
server samples when using the Zephyr Bluetooth Low Energy
Controller, i.e. Tx ISO SDUs dropped due to delayed ISO data
send calls in upper layer with respect to payload number in
the current ISO radio events in the Controller.

Add Zephyr Bluetooth Low Energy Controller overlay config for
the unicast audio client and server samples.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: unicast audio client and server overlay config

Add Zephyr Bluetooth Low Energy Controller overlay config for
the unicast audio client and server samples.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix Tx ISO SDUs dropped due to strict SN check

Introduce a Kconfig BT_CTLR_ISOAL_SN_STRICT to relax the Tx
ISO Data SDUs being dropped due to strict check of sequence
numbers in ISOAL, i.e. dropped when ISO Data SDUs are
delayed from upper layer with respect to the current
payload number in the ISO radio events.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: Relax Tx ISO SDUs dropped in unicast audio samples

Relax Tx ISO SDUs being dropped in unicast audio client and
server samples when using the Zephyr Bluetooth Low Energy
Controller, i.e. Tx ISO SDUs dropped due to delayed ISO data
send calls in upper layer with respect to payload number in
the current ISO radio events in the Controller.

This option is disabled in the sample and users can use it
when need be.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@niag-Oticon Hope the new Kconfig is acceptable as a workaround for applications that is not able keep the timings and provide stale ISO data (sequence numbers)

Thalley commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -377,6 +377,19 @@ config BT_CTLR_SCAN_ENABLE_STRICT
 	  Enforce returning HCI Error Command Disallowed on enabling/disabling
 	  already enabled/disabled scanning.
 
+config BT_CTLR_ISOAL_SN_STRICT
+	bool "Enforce Strict Tx ISO Data Sequence Number use"
```
```suggestion
config BT_CTLR_ISOAL_PSN_STRICT
	bool "Enforce Strict Tx ISO Data Packet Sequence Number use"
```

Thalley replied:
Hmm, I see that the controller actually uses `sn` multiple places for this value, so perhaps we should just keep it as `SN`/`Sequence Number` rather than the full name of the field (Packet Sequence Number).

cvinayak replied:
Keeping "sequence number" phrase as ISOAL specification uses this phrase (and does not use "packet sequence number"). `Packet_Sequence_Number` is only mentioned in the context of HCI ISO Data packets.

Thalley replied:
Correct, but the sequence number in ISOAL is the 1-bit value that is either 0 or 1, right? 

This Kconfig only applies to the packet sequence number that the host provides when sending over HCI, right? 

I am OK with using use sequence number

cvinayak replied:
> but the sequence number in ISOAL is the 1-bit value that is either 0 or 1, right?

This is a bit confusing in Spec, in ISOAL SDU sequence number is 16-bit to match the value `Packet_Sequence_Number` in HCI ISO Data packet. The other `SN` and `NESN` these are in the PDU header.

```
The sequence number shall be incremented for each SDU_Interval, whether or
not an SDU was received from the upper layer or included in reports sent to the
upper layer.
```

The Spec uses the term `upper layer`, hence is open to having other interfaces (not just HCI as the interface) for sending/receiving Tx ISO Data packets.

Thalley replied:
Okay so "sequence number" in ISOAL refer to any sequence number from upper layers (does that include vendor specific data paths?), and "packet sequence number" is only used for HCI. 

I'd argue that if this check is only for HCI ISO data packets, then it makes sense to use the term "packet sequence number", but if it is for any upper layer, then stick with "sequence number".

Thalley commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -377,6 +377,19 @@ config BT_CTLR_SCAN_ENABLE_STRICT
+	bool "Enforce Strict Tx ISO Data Sequence Number use"
+	depends on BT_CTLR_ADV_ISO || BT_CTLR_CONN_ISO
+	default y
+	help
+	  Enforce dropping stale Tx ISO Data SDUs from upper layer by checking
+	  the next expected Sequence Number maintained in ISOAL.
+
+	  This option may be used to relax the Tx ISO Data SDUs being dropped
+	  due to strict check of sequence numbers in ISOAL, i.e. dropped when
+	  ISO Data SDUs are delayed from upper layer with respect to the current
+	  payload number in the ISO radio events.
```
Maybe it should be mentioned that this only applies for ISO over HCI?

cvinayak replied:
Done

niag-Oticon reviewed (commented):
Changes appear alright as long as the impact is understood.

niag-Oticon reviewed (changes requested):
Changes appear alright as long as the impact is understood.
I have some concerns that the current description does not fully capture the impact.
Also this should not be disabled for any qualification.

cvinayak committed:
samples: Bluetooth: Fix Tx ISO Data packet sequence number

Fix Tx ISO Data packet sequence number increment based on
the send timer timeout value.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Update BT_CTLR_ISOAL_SN_STRICT help text

Update BT_CTLR_ISOAL_SN_STRICT help text.

Co-authored-by: Nirosharn Amarasinghe <niag@demant.com>
Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

niag-Oticon reviewed (dismissed):
Changes appear alright as long as the impact is understood.
This should not be disabled for any qualification.

Thalley commented on samples/bluetooth/unicast_audio_client/src/main.c:
```diff
@@ -58,11 +58,23 @@ static K_SEM_DEFINE(sem_stream_qos, 0, ARRAY_SIZE(sinks) + ARRAY_SIZE(sources));
-			return sinks[i].seq_num++;
+			seq_num = sinks[i].seq_num;
+
+#if defined(CONFIG_LIBLC3)
+			sinks[i].seq_num++;
+#else
+			sinks[i].seq_num += (AUDIO_DATA_TIMEOUT_US /
+					     codec_configuration.qos.interval);
+#endif
```
```suggestion
			if (IS_ENABLED(CONFIG_LIBLC3)) {
				sinks[i].seq_num++;
			} else {
				sinks[i].seq_num += (AUDIO_DATA_TIMEOUT_US /
						     codec_configuration.qos.interval);	
			}
```

cvinayak replied:
Done

MariuszSkamra commented:
@cvinayak Why does the controller actually check the SeqNum and drop the packets instead of just sending it? Controller and Host are running in different clock domains, and there's always a drift.
So the controller cannot rely the SeqNum. I think that if there is packet on time, it shall be sent, if not ISO Null PDU shall be sent.

Thalley commented on samples/bluetooth/unicast_audio_server/src/main.c:
```diff
@@ -69,11 +69,24 @@ static const struct bt_data ad[] = {
-			return source_streams[i].seq_num++;
+			seq_num = source_streams[i].seq_num;
+
+#if defined(CONFIG_LIBLC3)
+			source_streams[i].seq_num++;
+#else
+			source_streams[i].seq_num += (AUDIO_DATA_TIMEOUT_US /
+						      SDU_INTERVAL_US);
+#endif
```
```suggestion
			if (IS_ENABLED(CONFIG_LIBLC3)) {
				sinks[i].seq_num++;
			} else {
				sinks[i].seq_num += (AUDIO_DATA_TIMEOUT_US /
						     codec_configuration.qos.interval);	
			}
```

Thalley replied:
Alternatively you can also do
```suggestion
			if (IS_ENABLED(CONFIG_LIBLC3)) {
				return sinks[i].seq_num++;
			} else {
				return sinks[i].seq_num += (AUDIO_DATA_TIMEOUT_US /
							    codec_configuration.qos.interval);	
			}
```

And remove the local variable again

cvinayak replied:
Updated.

Thalley commented on samples/bluetooth/unicast_audio_server/src/main.c:
```diff
@@ -69,11 +69,24 @@ static const struct bt_data ad[] = {
+
 static uint16_t get_and_incr_seq_num(const struct bt_bap_stream *stream)
 {
+	uint16_t seq_num;
```
See comment below, but in any case this can be move to the inner scope where it is used and returned

cvinayak replied:
Done

Thalley commented on samples/bluetooth/unicast_audio_server/src/main.c:
```diff
@@ -214,7 +227,11 @@ static void audio_timer_timeout(struct k_work *work)
 	}
 
-	k_work_schedule(&audio_send_work, K_MSEC(1000U));
+#if defined(CONFIG_LIBLC3)
+	k_work_schedule(&audio_send_work, K_USEC(MAX_FRAME_DURATION_US));
+#else
+	k_work_schedule(&audio_send_work, K_USEC(AUDIO_DATA_TIMEOUT_US));
+#endif
```
Ditto with using `IS_ENABLED`. 

If we want to do this correctly, we should use the `bt_codec_qos.interval` from the `lc3_qos` callback as the interval here. 

We already store the `max_sdu` size, so should be easy to store the interval as well

cvinayak replied:
I am not familiar with the implementation of codec, if you provide suggested code I can include it in this PR. Or you are welcome to send a follow up PR.

No change done, as `MAX_FRAME_DURATION_US` is currently conditionally compiled hence not able to use `IS_ENABLED`

cvinayak commented:
> @cvinayak Why does the controller actually check the SeqNum and drop the packets instead of just sending it? If controller and Host are running in different clock domains, and there's always a drift. So the controller cannot rely the SeqNum. I think that if there is packet on time, it shall be sent, if not ISO Null PDU shall be sent.

If the SDUs are not dropped, as @niag-Oticon mentions "creating skews in the received stream of SDUs" 

In the Spec:
```
The sequence number shall be incremented for each SDU_Interval, whether or
not an SDU was received from the upper layer or included in reports sent to the
upper layer.

The sequence number shall be 16 bits and shall be included in the
Packet_Sequence_Number parameter in HCI ISO Data Packets
```

Also, if the implement is known to drift then framed PDUs have to be used.

```
Framed PDUs shall be used when the requirements for using unframed PDUs
are not met. Framed PDUs support the aggregation of data from multiple SDUs
into a single PDU. The maximum allowed drift (MaxDrift) on the average timing
of SDU delivery shall not exceed 100 ppm from the configured value of
SDU_Interval.
```

cvinayak commented:
@Thalley rebased on resolve conflict, request your re-approval.
